### PR TITLE
(Re)Enable Cholesky factorization of matrices

### PR DESCRIPTION
**Pull request by benson31**
Calls legacy code on CPU; calls cuSOLVER for local GPU factorizations.

TODO:

- [x] Complete rocSolver integration for the ROCm stack.
- [x] Test GPU code.

EDIT: We have decided that adding straightforward LAPACK POTRF bindings for CPU matrices when we have LAPACK support is best left for future work.

### CODE REVIEW

**Review by oyamay — APPROVED** [review]
It looks good to me. At least `El::Cholesky` and `El::Trsm` work correctly on LBANN if cuSolver is linked manually.

File: include/hydrogen/blas/BLAS_Common.hpp [hunk 1]
```diff
@@ -77,6 +121,12 @@ enum class SideMode
     RIGHT,
 };
 
+enum class DiagType
```
Comment (timmoon10) [inline include/hydrogen/blas/BLAS_Common.hpp:124]: It seems like we're duplicating enums here and in `core/types.hpp` (see also `SideMode`/`LeftOrRight`). If you prefer keeping the enums in this header, is it feasible to make the `types.hpp` enums alias to these enums?

File: include/hydrogen/blas/GPU_BLAS_impl.hpp [hunk 2]
```diff
@@ -274,6 +280,82 @@ void GemvImpl(
...
+        ToNativeFillMode(uplo), ToNativeTransposeMode(trans),
+        ToSizeT(n), ToSizeT(k),
+        alpha,
+        reinterpret_cast<CNTP>(A), ToSizeT(lda),
+        beta,
+        reinterpret_cast<NTP>(C), ToSizeT(ldc));
+}
+
+template <typename T, typename SizeT,
+          typename=EnableWhen<IsSupportedType<T, BLAS_Op::SYRK>>>
```
Comment (timmoon10) [inline include/hydrogen/blas/GPU_BLAS_impl.hpp]:
```suggestion
          typename=EnableWhen<IsSupportedType<T, BLAS_Op::TRSM>>>
```

File: include/hydrogen/blas/GPU_BLAS_impl.hpp [hunk 3]
```diff
@@ -536,6 +618,60 @@ void GemvImpl(
...
+    SyncInfo<Device::GPU> const&)
+{
+    std::ostringstream oss;
+    oss << "No valid implementation of SYRK for T="
+        << TypeTraits<T>::Name();
+    throw std::logic_error(oss.str());
+}
+
+template <typename T, typename SizeT,
+          typename=EnableUnless<IsSupportedType<T, BLAS_Op::SYRK>>,
```
Comment (timmoon10) [inline include/hydrogen/blas/GPU_BLAS_impl.hpp]:
```suggestion
          typename=EnableUnless<IsSupportedType<T, BLAS_Op::TRSM>>,
```

File: include/hydrogen/blas/GPU_BLAS_impl.hpp [hunk 4]
```diff
@@ -727,5 +911,167 @@ void Dgmm(SideMode side,
...
+    simple_buffer<gpu_lapack_impl::InfoT, Device::GPU> info(1, si);
+    CholeskyFactorizeImpl(uplo, n, A, lda,
+                          workspace, workspace_size, info.data(), si);
+#ifndef EL_RELEASE
+    gpu_blas_impl::InfoT host_info;
+    Copy1DToHost(info.data(), &host_info, 1, si);
+    Synchronize(si);
+    if (host_info > gpu_blas_impl::InfoT(0))
+        throw std::runtime_error("Cholesky: Matrix not HPD.");
+    else if (host_inf < gpu::blas_impl::InfoT(0))
```
Comment (timmoon10) [inline include/hydrogen/blas/GPU_BLAS_impl.hpp]:
```suggestion
    else if (host_info < gpu::blas_impl::InfoT(0))
```

File: src/hydrogen/device/rocSOLVER.cpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+void InitializeDense(rocblas_handle handle)
+{
+    rocblas::Initialize(handle);
+    rocsolver_is_initialized_ = true;
+}
+
+void FinalizeDense()
+{
+    rocblas::Finalize();
```
Comment (timmoon10) [inline src/hydrogen/device/rocSOLVER.cpp:30]:
```suggestion
    rocblas::Finalize();
    rocsolver_is_initialized_ = false;
```

**Review by timmoon10 — APPROVED** [review]
LGTM

File: include/hydrogen/blas/BLAS_Common.hpp (hunk 1 above)
Comment (benson31) [reply include/hydrogen/blas/BLAS_Common.hpp:124]: Yes; will address in a future PR.

**Review by bvanessen — APPROVED** [review]

**Pull request merged**